File: src/lyxserver.ts
/**
 * LyXServer client module.
 *
 * Communicates with a running LyX instance to send LFUN commands and
 * read responses. Supports two transport mechanisms:
 *
 *   Linux/macOS: Unix domain socket (ServerSocket.cpp)
 *     - Protocol: LYXCMD:<lfung> <args>
 *     - Response: INFO:<cmd>:<msg> or ERROR:<cmd>:<msg>
 *     - Discovery: $LYXSOCKET env var, or scan /tmp/lyx_tmpdir*
 *
 *   Windows: Named pipes (Server.cpp)
 *     - Protocol: LYXCMD:<client>:<lfung> <args>
 *     - Response: INFO:<client>:<lfung>:<msg> or ERROR:...
 *     - Discovery: default %APPDATA%\LyX2.5\lyxpipe
 *     - Pipe paths: \\.\pipe\<base>.in (write), \\.\pipe\<base>.out (read)
 */

import * as path from "@std/path";

/**
 * Race a promise against a timeout. The timer is cleared once the race settles
 * (either way) so it doesn't keep the process alive after the work completes
 * (test_report_34 Finding 8). Note (dev log 75): on Windows a pending
 * Deno.open/read on a named pipe cannot be cancelled at the OS level, so a
 * timed-out open may still be lingering — but the caller can degrade gracefully
 * because this promise rejects when the timeout wins.
 */
async function withTimeout<T>(promise: Promise<T>, ms: number): Promise<T> {
  let timer: ReturnType<typeof setTimeout> | null = null;
  try {
    return await Promise.race([
      promise,
      new Promise<never>((_, reject) => {
        timer = setTimeout(() => reject(new Error(`timeout after ${ms}ms`)), ms);
      }),
    ]);
  } finally {
    if (timer !== null) clearTimeout(timer);
  }
}

// ---------------------------------------------------------------------------
// Pipe / socket discovery
// ---------------------------------------------------------------------------

/**
 * Discover the LyX pipe path on Windows.
 * Returns the base pipe path (without .in/.out suffix), or null if not found.
 */
function discoverWindowsPipePath(): string | null {
  // 1. Check LYXSOCKET environment variable
  const envSocket = Deno.env.get("LYXSOCKET");
  if (envSocket) return envSocket;

  // 2. Default location: %APPDATA%\LyX2.5\lyxpipe
  // Only return a path if LyX is actually running (check via tasklist).
  // Otherwise, Deno.open() on \\.\pipe\... blocks indefinitely waiting
  // for a pipe server that never comes.
  const appData = Deno.env.get("APPDATA");
  if (!appData) return null;

  const lyxDir = path.join(appData, "LyX2.5");
  try {
    Deno.statSync(lyxDir);
  } catch {
    return null;
  }

  // Quick check: is LyX running? If not, the pipe won't be listening.
  if (!isLyxRunning()) return null;

  return path.join(lyxDir, "lyxpipe");
}

/**
 * Check if a LyX process is running on Windows.
 * Uses tasklist.exe to avoid blocking on pipe CreateFile.
 */
function isLyxRunning(): boolean {
  try {
    const cmd = new Deno.Command("tasklist", {
      args: ["/fi", "IMAGENAME eq LyX.exe", "/nh"],
      stdout: "piped",
      stderr: "null",
    });
    const output = new TextDecoder().decode(cmd.outputSync().stdout);
    return output.includes("LyX.exe");
  } catch {
    return false;
  }
}

/**
 * Discover the LyX socket path on Unix.
 * Returns the socket path, or null if not found.
 */
function discoverUnixSocket(): string | null {
  // 1. Check LYXSOCKET environment variable
  const envSocket = Deno.env.get("LYXSOCKET");
  if (envSocket) return envSocket;

  // 2. Scan temp directories for lyxsocket files
  try {
    const tmpDir = Deno.env.get("TMPDIR") || "/tmp";
    for (const entry of Deno.readDirSync(tmpDir)) {
      if (entry.isDirectory && entry.name.startsWith("lyx_tmpdir")) {
        const socketPath = path.join(tmpDir, entry.name, "lyxsocket");
        try {
          const stat = Deno.statSync(socketPath);
          if (stat.isFile || stat.isSocket) return socketPath;
        } catch {
          // Socket doesn't exist in this temp dir
        }
      }
    }
  } catch {
    // Can't read temp dir
  }

  return null;
}

// ---------------------------------------------------------------------------
// Unix socket transport
// ---------------------------------------------------------------------------

async function sendViaUnixSocket(
  socketPath: string,
  lfuns: string[],
): Promise<{ sent: boolean; confirmed: boolean; errored: boolean }> {
  let conn: Deno.Conn | null = null;
  try {
    // Timeout: Deno.connect on a Unix socket hangs if the socket file exists
    // but no process accepts. Bounded via withTimeout (timer cleared on settle —
    // test_report_34 Finding 8; S1 dedup of the triplicated race pattern).
    conn = await withTimeout(Deno.connect({ path: socketPath, transport: "unix" }), 5000);
  } catch {
    return { sent: false, confirmed: false, errored: false };
  }

  const encoder = new TextEncoder();
  const decoder = new TextDecoder();
  const buf = new Uint8Array(4096);

  let sent = false;
  let confirmed = false;
  let errored = false;

  async function readLine(): Promise<string | null> {
    let data = "";
    while (true) {
      // 5-second timeout per read — protects against a hung server
      // (bounded via withTimeout; timer cleared on settle). The .catch guards
      // against an unhandled rejection if a timed-out read later errors when
      // the connection is closed.
      const n = await withTimeout(conn!.read(buf).catch(() => null), 5000);
      if (n === null) return data || null;
      data += decoder.decode(buf.subarray(0, n));
      const nl = data.indexOf("\n");
      if (nl !== -1) {
        const line = data.substring(0, nl);
        data = data.substring(nl + 1);
        return line;
      }
    }
  }

  async function sendLine(line: string): Promise<void> {
    await conn!.write(encoder.encode(line + "\n"));
  }

  try {
    // Handshake
    await sendLine("HELLO:");
    const helloResp = await readLine();
    if (!helloResp || !helloResp.startsWith("HELLO:")) {
      try { conn.close(); } catch { /* ignore */ }
      return { sent: false, confirmed: false, errored: false };
    }

    for (const lfun of lfuns) {
      await sendLine(`LYXCMD:${lfun}`);
      sent = true;
      const resp = await readLine();
      if (!resp) {
        try { conn.close(); } catch { /* ignore */ }
        return { sent, confirmed: false, errored };
      }
      // "Command disabled" means the LFUN isn't needed (e.g. buffer-write
      // when there are no unsaved changes). Treat as success.
      if (resp.startsWith("ERROR:") && !resp.includes("Command disabled")) errored = true;
    }
    // Every command produced a response — the whole sequence is confirmed.
    confirmed = true;

    await sendLine("BYE:");
    try { conn.close(); } catch { /* ignore */ }
    return { sent, confirmed, errored };
  } catch {
    try { conn.close(); } catch { /* ignore */ }
    return { sent, confirmed, errored };
  }
}

// ---------------------------------------------------------------------------
// Windows named pipe transport
// ---------------------------------------------------------------------------

/**
 * Build a full Windows named pipe path from a base path.
 * LyX uses \\.\pipe\<full_filesystem_path_with_backslashes>.in / .out
 */
function winPipePath(basePath: string, suffix: ".in" | ".out"): string {
  // Convert forward slashes to backslashes for the pipe namespace
  const winPath = basePath.replace(/\//g, "\\");
  return `\\\\.\\pipe\\${winPath}${suffix}`;
}

async function sendViaNamedPipe(
  pipeBase: string,
  lfuns: string[],
): Promise<{ sent: boolean; confirmed: boolean; errored: boolean }> {
  const inPipe = winPipePath(pipeBase, ".in");
  const outPipe = winPipePath(pipeBase, ".out");

  const encoder = new TextEncoder();
  const decoder = new TextDecoder();
  const readBuf = new Uint8Array(512);

  // test_report_38 F1 — Server.cpp delivery race: the pipe loop only copies the
  // shared outbuf_ to .out instances when it wakes *after* a reply exists. A
  // .out reader that connects *after* the command is written can consume the
  // wake before the reply is ready, so the reply lands on the *next* connection
  // where the client-name filter discards it → ~50% of responses lost. Verified
  // empirically (2026-08-01) that no client read strategy fixes this reliably —
  // reader-before-reply and keep-open both make it worse (7/8 lost) and raced
  // reads leave the process lingering (dev log 75). So lq is robust instead:
  //   - `sent` — the command was written to .in (LyX accepted and will execute
  //     it; the operation is NOT lost even when the response is).
  //   - `confirmed` — we also received the response (best-effort on Windows).
  //   - `errored` — a confirmed response was a real ERROR (not "Command disabled").
  // Callers treat `sent && !confirmed` as "attempted, unconfirmed" (proceed with
  // a warning) rather than as a failed operation.
  let sent = false;
  let confirmed = false;
  let errored = false;

  async function sendOnce(): Promise<void> {
    // Fresh client name per attempt so a stale response from a previous attempt
    // (flushed by the shared outbuf_ into every .out) is never misattributed.
    const clientName = `lq${Date.now()}`;

    let inFile: Deno.FsFile | null = null;
    try {
      // Timeout: Deno.open on a Windows named pipe blocks indefinitely if no
      // pipe server is listening (CreateFile behavior). Bounded via withTimeout
      // (timer cleared on settle — test_report_34 Finding 8).
      inFile = await withTimeout(Deno.open(inPipe, { write: true }), 5000);
    } catch {
      return; // couldn't even open .in — genuine disconnect, nothing dispatched
    }

    try {
      for (const lfun of lfuns) {
        try {
          // Send command to .in (keep .in open — closing it ends the session)
          await inFile.write(encoder.encode(`LYXCMD:${clientName}:${lfun}\n`));
        } catch {
          // Write failed after a successful open — server likely died mid-call.
          return;
        }
        sent = true;

        // Poll .out: the server needs time to process the command and write the
        // response. Each open-drain-close is a fresh connection whose connect
        // wakes the server loop, which is when it flushes outbuf_ to the .out
        // instances. A lost response surfaces as a later retry of the sequence.
        let response: string | null = null;
        for (const delay of [50, 100, 200, 500, 1000]) {
          await new Promise(r => setTimeout(r, delay));
          response = await tryReadResponse(outPipe, readBuf, decoder, clientName);
          if (response !== null) break;
        }

        if (response === null) return; // incomplete attempt — retry the sequence
        // "Command disabled" means the LFUN isn't needed (e.g. buffer-write
        // when there are no unsaved changes). Treat as success.
        if (response.startsWith("ERROR:") && !response.includes("Command disabled")) errored = true;
      }
      // Every command produced a response — the whole sequence is confirmed.
      confirmed = true;
    } finally {
      try { inFile.close(); } catch { /* ignore */ }
    }
  }

  for (let attempt = 0; attempt < 3 && !confirmed; attempt++) {
    await sendOnce();
  }
  return { sent, confirmed, errored };
}

/**
 * Filter a chunk of .out data down to the last response line addressed to
 * `clientName`. LyX's server accumulates responses for ALL clients in one
 * shared buffer and copies it to every .out instance (Server.cpp
 * WRITING_STATE), so lines must be correlated by client name, not by position
 * (test_report_34 Finding 9). Returns null when no matching line is present.
 */
export function filterResponses(data: string, clientName: string): string | null {
  const lines = data.split("\n").filter(l => l);
  const mine = lines.filter(l =>
    l.startsWith(`INFO:${clientName}:`) || l.startsWith(`ERROR:${clientName}:`));
  return mine.length > 0 ? mine[mine.length - 1] : null;
}

/**
 * Open .out, drain all currently available data, and return the last response
 * line addressed to `clientName` (or null if none arrived yet). Each call is a
 * fresh connection whose connect wakes the Server.cpp loop — which is when it
 * flushes its shared outbuf_ to the .out instances (test_report_38 F1).
 *
 * Draining is essential: without it, stale responses from other clients would
 * persist and be misattributed by a later connection (false REFRESH_PRE_ERROR
 * / false success — test_report_34 Finding 9). Reads here are blocking (until
 * EOF/error, which the server triggers by resetting the instance) — never raced
 * with a timeout, since an abandoned read on a Windows pipe pins the process
 * (dev log 75).
 */
async function tryReadResponse(
  outPipe: string,
  buf: Uint8Array,
  decoder: TextDecoder,
  clientName: string,
): Promise<string | null> {
  let outFile: Deno.FsFile | null = null;
  try {
    outFile = await withTimeout(Deno.open(outPipe, { read: true }), 2000);
  } catch {
    return null;
  }

  try {
    let data = "";
    // Drain everything currently available so stale responses from other
    // clients don't persist into the next connection.
    for (let attempt = 0; attempt < 40; attempt++) {
      try {
        const n = await outFile.read(buf);
        if (n === null || n === 0) break;
        data += decoder.decode(buf.subarray(0, n));
      } catch (e) {
        if (e instanceof Deno.errors.BrokenPipe ||
            (e instanceof Error && e.message.includes("os error 233"))) {
          break; // Pipe disconnected
        }
        throw e;
      }
      await new Promise(r => setTimeout(r, 30));
    }

    return filterResponses(data, clientName);
  } finally {
    try { outFile.close(); } catch { /* ignore */ }
  }
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Connect to a running LyX instance and send a sequence of LFUN commands.
 *
 * Returns `{ sent, confirmed, errored }`:
 * - `sent` — the command was dispatched to LyX (written to the pipe/socket).
 *   On Windows this is the reliable part: Server.cpp executes the LFUN even
 *   when the response is lost to the delivery race (test_report_38 F1).
 * - `confirmed` — every command also produced a response (INFO or ERROR).
 *   This is best-effort on Windows (see F1); a false `confirmed` with a true
 *   `sent` means the operation ran but the confirmation was lost.
 * - `errored` — a confirmed response was a real ERROR (not "Command disabled").
 *
 * @param lfuns - Array of LFUN command strings (e.g., ["buffer-switch /path/file.lyx", "buffer-reload"])
 */
export async function sendLyxCommands(
  lfuns: string[],
): Promise<{ sent: boolean; confirmed: boolean; errored: boolean }> {
  try {
    if (Deno.build.os === "windows") {
      const pipeBase = discoverWindowsPipePath();
      if (!pipeBase) return { sent: false, confirmed: false, errored: false };
      return await sendViaNamedPipe(pipeBase, lfuns);
    } else {
      const socketPath = discoverUnixSocket();
      if (!socketPath) return { sent: false, confirmed: false, errored: false };
      return await sendViaUnixSocket(socketPath, lfuns);
    }
  } catch {
    // Timeout or other transport error — LyX is unreachable.
    // Caller should degrade gracefully (mutation was already committed).
    return { sent: false, confirmed: false, errored: false };
  }
}

/**
 * Check if a running LyX instance with LyXServer enabled is reachable.
 * Used at `lq init` time to verify configuration.
 */
export function checkLyxServerAvailable(): boolean {
  if (Deno.build.os === "windows") {
    return discoverWindowsPipePath() !== null;
  }
  return discoverUnixSocket() !== null;
}

File: src/lyxserver.rs
//! LyXServer client (Deno `lyxserver.ts`). Blocking I/O; no tokio (C5).

use std::fs;
use std::io::{self, Read, Write};
#[cfg(unix)]
use std::path::Path;
#[cfg(windows)]
use std::path::PathBuf;
#[cfg(windows)]
use std::thread;
use std::time::Duration;
#[cfg(windows)]
use std::time::{SystemTime, UNIX_EPOCH};

/// DL87 Option A: dispatch (`sent`) is reliable; a missing reply is unconfirmed, not failure.
#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
pub struct SendResult {
    pub sent: bool,
    pub confirmed: bool,
    pub errored: bool,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum RefreshMode {
    None,
    Reload,
    SaveReload,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum RefreshPreStep {
    Ok,
    Disconnect,
    Unconfirmed,
    Error,
}

/// First space becomes func/arg; arg runs to `\n` so `C:` stays intact (DL128).
pub fn build_pipe_command(client_name: &str, lfun: &str) -> String {
    let wire = match lfun.find(' ') {
        None => lfun.to_string(),
        Some(sep) => format!("{}:{}", &lfun[..sep], &lfun[sep + 1..]),
    };
    format!("LYXCMD:{client_name}:{wire}\n")
}

/// Last `INFO|ERROR:<client>:` line. Exact client name (DL82); not a prefix.
pub fn filter_responses(data: &str, client_name: &str) -> Option<String> {
    let info = format!("INFO:{client_name}:");
    let error = format!("ERROR:{client_name}:");
    data.split('\n')
        .filter(|line| !line.is_empty())
        .rfind(|line| line.starts_with(&info) || line.starts_with(&error))
        .map(str::to_string)
}

pub(crate) fn send_lyx_commands(lfuns: &[&str]) -> SendResult {
    #[cfg(windows)]
    {
        match discover_windows_pipe_path() {
            Some(base) => send_via_named_pipe(&base, lfuns),
            None => SendResult::default(),
        }
    }
    #[cfg(unix)]
    {
        match discover_unix_socket() {
            Some(path) => send_via_unix_socket(&path, lfuns),
            None => SendResult::default(),
        }
    }
    #[cfg(not(any(windows, unix)))]
    {
        let _ = lfuns;
        SendResult::default()
    }
}

pub(crate) fn check_lyx_server_available() -> bool {
    #[cfg(windows)]
    {
        match discover_windows_pipe_path() {
            Some(base) => probe_named_pipe(&base),
            None => false,
        }
    }
    #[cfg(unix)]
    {
        match discover_unix_socket() {
            Some(path) => send_via_unix_socket(&path, &["server-get-filename"]).sent,
            None => false,
        }
    }
    #[cfg(not(any(windows, unix)))]
    {
        false
    }
}

pub fn refresh_pre_step(file_path: &str, mode: RefreshMode) -> RefreshPreStep {
    if mode != RefreshMode::SaveReload {
        return RefreshPreStep::Ok;
    }
    let switch = format!("buffer-switch {}", absolute_file(file_path));
    let result = send_lyx_commands(&[switch.as_str(), "buffer-write"]);
    if !result.sent {
        RefreshPreStep::Disconnect
    } else if !result.confirmed {
        RefreshPreStep::Unconfirmed
    } else if result.errored {
        RefreshPreStep::Error
    } else {
        RefreshPreStep::Ok
    }
}

pub(crate) fn refresh_post_step(file_path: &str, mode: RefreshMode) -> Option<SendResult> {
    if mode == RefreshMode::None {
        return None;
    }
    let switch = format!("buffer-switch {}", absolute_file(file_path));
    Some(send_lyx_commands(&[switch.as_str(), "buffer-reload"]))
}

fn absolute_file(path: &str) -> String {
    std::path::absolute(path)
        .map(|p| p.to_string_lossy().into_owned())
        .unwrap_or_else(|_| path.to_string())
}

#[cfg(windows)]
fn now_millis() -> u128 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_millis())
        .unwrap_or(0)
}

fn env_nonempty(name: &str) -> Option<String> {
    match std::env::var(name) {
        Ok(value) if !value.is_empty() => Some(value),
        _ => None,
    }
}

fn is_error_response(resp: &str) -> bool {
    resp.starts_with("ERROR:") && !resp.contains("Command disabled")
}

// ---------------------------------------------------------------------------
// Windows named pipes
// ---------------------------------------------------------------------------

#[cfg(windows)]
fn discover_windows_pipe_path() -> Option<String> {
    if let Some(socket) = env_nonempty("LYXSOCKET") {
        return Some(socket);
    }
    let app_data = env_nonempty("APPDATA")?;
    let lyx_dir = PathBuf::from(app_data).join("LyX2.5");
    match fs::metadata(&lyx_dir) {
        Ok(meta) if meta.is_dir() => {}
        _ => return None,
    }
    if !is_lyx_running() {
        return None;
    }
    Some(lyx_dir.join("lyxpipe").to_string_lossy().into_owned())
}

#[cfg(windows)]
fn is_lyx_running() -> bool {
    use std::process::{Command, Stdio};
    let output = Command::new("tasklist")
        .args(["/fi", "IMAGENAME eq LyX.exe", "/nh"])
        .stdout(Stdio::piped())
        .stderr(Stdio::null())
        .output();
    match output {
        Ok(out) => String::from_utf8_lossy(&out.stdout).contains("LyX.exe"),
        Err(_) => false,
    }
}

#[cfg(windows)]
fn win_pipe_path(base: &str, suffix: &str) -> String {
    let win_path = base.replace('/', "\\");
    format!(r"\\.\pipe\{win_path}{suffix}")
}

/// Wait until a pipe instance is connectable, then open with `std`.
/// `WaitNamedPipeW` is the timeout; the thread does not linger (JC2).
#[cfg(windows)]
fn open_named_pipe(path: &str, write: bool, timeout_ms: u32) -> io::Result<fs::File> {
    let wide: Vec<u16> = path.encode_utf16().chain(std::iter::once(0)).collect();
    // SAFETY: `wide` is a NUL-terminated UTF-16 pipe name; WaitNamedPipeW only
    // reads it for the duration of the call.
    let ready =
        unsafe { windows_sys::Win32::System::Pipes::WaitNamedPipeW(wide.as_ptr(), timeout_ms) };
    if ready == 0 {
        return Err(io::Error::last_os_error());
    }
    let mut opts = fs::OpenOptions::new();
    if write {
        opts.write(true);
    } else {
        opts.read(true);
    }
    opts.open(path)
}

#[cfg(windows)]
fn send_via_named_pipe(pipe_base: &str, lfuns: &[&str]) -> SendResult {
    let mut result = SendResult::default();
    for _ in 0..3 {
        if result.confirmed {
            break;
        }
        send_named_pipe_once(pipe_base, lfuns, &mut result);
    }
    result
}

#[cfg(windows)]
fn send_named_pipe_once(pipe_base: &str, lfuns: &[&str], result: &mut SendResult) {
    let in_pipe = win_pipe_path(pipe_base, ".in");
    let out_pipe = win_pipe_path(pipe_base, ".out");
    let client_name = format!("lq{}", now_millis());
    let mut in_file = match open_named_pipe(&in_pipe, true, 5000) {
        Ok(file) => file,
        Err(_) => return,
    };

    for lfun in lfuns {
        let line = build_pipe_command(&client_name, lfun);
        if in_file.write_all(line.as_bytes()).is_err() {
            return;
        }
        result.sent = true;

        let mut response = None;
        for delay in [50, 100, 200, 500, 1000] {
            thread::sleep(Duration::from_millis(delay));
            response = try_read_response(&out_pipe, &client_name);
            if response.is_some() {
                break;
            }
        }
        let Some(resp) = response else {
            return;
        };
        if is_error_response(&resp) {
            result.errored = true;
        }
    }
    result.confirmed = true;
}

#[cfg(windows)]
fn try_read_response(out_pipe: &str, client_name: &str) -> Option<String> {
    let mut out_file = match open_named_pipe(out_pipe, false, 2000) {
        Ok(file) => file,
        Err(_) => return None,
    };

    let mut data = String::new();
    let mut buf = [0u8; 512];
    for _ in 0..40 {
        match out_file.read(&mut buf) {
            Ok(0) => break,
            Ok(n) => data.push_str(&String::from_utf8_lossy(&buf[..n])),
            Err(err) => {
                if err.kind() == io::ErrorKind::BrokenPipe || err.raw_os_error() == Some(233) {
                    break;
                }
                break;
            }
        }
        thread::sleep(Duration::from_millis(30));
    }
    filter_responses(&data, client_name)
}

#[cfg(windows)]
fn probe_named_pipe(pipe_base: &str) -> bool {
    let in_pipe = win_pipe_path(pipe_base, ".in");
    let client_name = format!("lqprobe{}", now_millis());
    let mut in_file = match open_named_pipe(&in_pipe, true, 3000) {
        Ok(file) => file,
        Err(_) => return false,
    };
    let line = format!("LYXCMD:{client_name}:server-get-filename\n");
    in_file.write_all(line.as_bytes()).is_ok()
}

// ---------------------------------------------------------------------------
// Unix domain socket
// ---------------------------------------------------------------------------

#[cfg(unix)]
fn discover_unix_socket() -> Option<String> {
    if let Some(socket) = env_nonempty("LYXSOCKET") {
        return Some(socket);
    }
    let tmp = env_nonempty("TMPDIR").unwrap_or_else(|| "/tmp".into());
    let entries = fs::read_dir(&tmp).ok()?;
    for entry in entries.flatten() {
        let name = entry.file_name();
        let Some(name) = name.to_str() else {
            continue;
        };
        if !entry.file_type().ok().is_some_and(|t| t.is_dir()) || !name.starts_with("lyx_tmpdir") {
            continue;
        }
        let socket_path = Path::new(&tmp).join(name).join("lyxsocket");
        if let Ok(meta) = fs::metadata(&socket_path) {
            use std::os::unix::fs::FileTypeExt;
            let ft = meta.file_type();
            if ft.is_file() || ft.is_socket() {
                return Some(socket_path.to_string_lossy().into_owned());
            }
        }
    }
    None
}

#[cfg(unix)]
fn send_via_unix_socket(socket_path: &str, lfuns: &[&str]) -> SendResult {
    use std::os::unix::net::UnixStream;

    let stream = match UnixStream::connect(socket_path) {
        Ok(stream) => stream,
        Err(_) => return SendResult::default(),
    };
    if stream
        .set_read_timeout(Some(Duration::from_millis(5000)))
        .is_err()
        || stream
            .set_write_timeout(Some(Duration::from_millis(5000)))
            .is_err()
    {
        return SendResult::default();
    }

    let mut reader = LineReader {
        stream,
        leftover: String::new(),
    };
    if reader.send_line("HELLO:").is_err() {
        return SendResult::default();
    }
    match reader.read_line() {
        Ok(Some(hello)) if hello.starts_with("HELLO:") => {}
        _ => return SendResult::default(),
    }

    let mut result = SendResult::default();
    for lfun in lfuns {
        if reader.send_line(&format!("LYXCMD:{lfun}")).is_err() {
            return result;
        }
        result.sent = true;
        match reader.read_line() {
            Ok(Some(resp)) => {
                if is_error_response(&resp) {
                    result.errored = true;
                }
            }
            _ => return result,
        }
    }
    result.confirmed = true;
    let _ = reader.send_line("BYE:");
    result
}

#[cfg(unix)]
struct LineReader {
    stream: std::os::unix::net::UnixStream,
    leftover: String,
}

#[cfg(unix)]
impl LineReader {
    fn send_line(&mut self, line: &str) -> io::Result<()> {
        self.stream.write_all(line.as_bytes())?;
        self.stream.write_all(b"\n")
    }

    fn read_line(&mut self) -> io::Result<Option<String>> {
        let mut buf = [0u8; 4096];
        loop {
            if let Some(nl) = self.leftover.find('\n') {
                let line = self.leftover[..nl].to_string();
                self.leftover = self.leftover[nl + 1..].to_string();
                return Ok(Some(line));
            }
            match self.stream.read(&mut buf) {
                Ok(0) => {
                    if self.leftover.is_empty() {
                        return Ok(None);
                    }
                    let line = std::mem::take(&mut self.leftover);
                    return Ok(Some(line));
                }
                Ok(n) => self.leftover.push_str(&String::from_utf8_lossy(&buf[..n])),
                Err(err) if err.kind() == io::ErrorKind::TimedOut => return Ok(None),
                Err(err) => return Err(err),
            }
        }
    }
}
